File: packages/contracts/src/deflate.ts
/**
 * Raw DEFLATE (RFC 1951, no zlib/gzip wrapper) — the only codec permitted
 * inside WFP2/WFU1 envelopes per docs/06.
 *
 * Two encoders are provided:
 * - `deflateRaw`: real compression via fflate (browser + Node).
 * - `deflateRawStored`: emits valid DEFLATE "stored" (uncompressed) blocks.
 *   The WoW addon can produce stored blocks in pure Lua without a compression
 *   library, and any conforming inflater — zlib included — accepts them.
 *   The website may emit WFU1 bundles with stored blocks so the addon can
 *   inflate them without vendored deflate code.
 */

import { deflateSync, inflateSync } from "fflate";
import { ProtocolError } from "./errors.js";

export function deflateRaw(data: Uint8Array): Uint8Array {
  // fflate's deflate is a raw DEFLATE stream (no zlib/gzip wrapper).
  return deflateSync(data);
}

export function inflateRaw(data: Uint8Array, maxBytes: number): Uint8Array {
  try {
    const out = inflateSync(data);
    if (out.length > maxBytes) {
      throw new ProtocolError("decoded_too_large", `decompressed payload exceeds ${maxBytes} bytes`);
    }
    return out;
  } catch (e) {
    if (e instanceof ProtocolError) throw e;
    throw new ProtocolError("decompression_failed", "invalid raw-DEFLATE data");
  }
}

/** Emit a raw-DEFLATE stream made entirely of stored (BTYPE=00) blocks. */
export function deflateRawStored(data: Uint8Array): Uint8Array {
  const nBlocks = Math.max(1, Math.ceil(data.length / 65535));
  const out = new Uint8Array(data.length + nBlocks * 5);
  let o = 0;
  let i = 0;
  do {
    const remaining = data.length - i;
    const len = Math.min(remaining, 65535);
    const final = i + len >= data.length;
    out[o++] = final ? 0x01 : 0x00; // BFINAL + BTYPE=00 (already byte-aligned)
    out[o++] = len & 0xff;
    out[o++] = (len >> 8) & 0xff;
    out[o++] = ~len & 0xff;
    out[o++] = (~len >> 8) & 0xff;
    out.set(data.subarray(i, i + len), o);
    o += len;
    i += len;
  } while (i < data.length);
  return out.subarray(0, o);
}

/**
 * Inflate a raw-DEFLATE stream that uses only stored blocks. Used by tests and
 * mirrored by the addon's minimal Lua inflater. Rejects compressed blocks.
 */
export function inflateRawStored(data: Uint8Array, maxBytes: number): Uint8Array {
  const chunks: Uint8Array[] = [];
  let total = 0;
  let o = 0;
  for (;;) {
    if (o + 5 > data.length) {
      throw new ProtocolError("decompression_failed", "truncated stored block header");
    }
    const header = data[o]!;
    const bfinal = header & 1;
    const btype = (header >> 1) & 3;
    if (btype !== 0 || (header & 0xf8) !== 0) {
      throw new ProtocolError("decompression_failed", "not a stored DEFLATE block");
    }
    const len = data[o + 1]! | (data[o + 2]! << 8);
    const nlen = data[o + 3]! | (data[o + 4]! << 8);
    if (nlen !== (~len & 0xffff)) {
      throw new ProtocolError("decompression_failed", "stored block NLEN mismatch");
    }
    o += 5;
    if (o + len > data.length) {
      throw new ProtocolError("decompression_failed", "truncated stored block data");
    }
    chunks.push(data.subarray(o, o + len));
    total += len;
    if (total > maxBytes) {
      throw new ProtocolError("decoded_too_large", `decompressed payload exceeds ${maxBytes} bytes`);
    }
    o += len;
    if (bfinal) break;
  }
  const out = new Uint8Array(total);
  let p = 0;
  for (const c of chunks) {
    out.set(c, p);
    p += c.length;
  }
  return out;
}

File: apps/addon/AzerothCombatLeague/Util/Deflate.lua
--[[
  Util/Deflate.lua — raw DEFLATE (RFC 1951) using stored (BTYPE=00) blocks only.

  Stored blocks are valid DEFLATE: every conforming inflater (zlib, fflate,
  browser DecompressionStream) accepts them, so WFP2 exports produced here are
  readable everywhere. Decoding is likewise limited to stored blocks — WFU1
  bundles bound for the addon are emitted stored-only by the website. A stream
  containing fixed/dynamic Huffman blocks is rejected with
  "unsupported_compression" rather than misdecoded.
]]
local _, ACL = ...

local Deflate = {}

--- Encode data as a raw-DEFLATE stream of stored blocks. Returns a string.
function Deflate.stored(data)
  local n = #data
  local parts = {}
  local i = 1
  while true do
    local offset = i - 1
    local remaining = n - offset
    local len = math.min(remaining, 65535)
    local final = offset + len >= n
    local nlen = 65535 - len -- ~len & 0xffff
    parts[#parts + 1] = string.char(
      final and 1 or 0,
      len % 256, math.floor(len / 256) % 256,
      nlen % 256, math.floor(nlen / 256) % 256)
    if len > 0 then
      parts[#parts + 1] = data:sub(i, i + len - 1)
    end
    i = i + len
    if final then break end
  end
  return table.concat(parts)
end

--- Inflate a stored-block-only raw-DEFLATE stream.
--- Returns string, or nil + error code.
function Deflate.inflateStored(data, maxBytes)
  maxBytes = maxBytes or (256 * 1024)
  local n = #data
  local out = {}
  local total = 0
  local o = 1
  while true do
    if o + 4 > n then return nil, "decompression_failed" end
    local header = data:byte(o)
    -- stored blocks are byte-aligned: header must be exactly 0x00 or 0x01
    if header ~= 0 and header ~= 1 then
      return nil, "unsupported_compression"
    end
    local bfinal = header
    local len = data:byte(o + 1) + data:byte(o + 2) * 256
    local nlen = data:byte(o + 3) + data:byte(o + 4) * 256
    if nlen ~= 65535 - len then
      return nil, "decompression_failed"
    end
    o = o + 5
    if o + len - 1 > n then return nil, "decompression_failed" end
    if len > 0 then
      out[#out + 1] = data:sub(o, o + len - 1)
      total = total + len
      if total > maxBytes then return nil, "decoded_too_large" end
    end
    o = o + len
    if bfinal == 1 then break end
  end
  return table.concat(out)
end

ACL.Deflate = Deflate
